File: 2D Array/spiralMatrix.java
public class spiralMatrix {
    public static void main(String[] args) {
         int matrix[][] = {{1,2,3,4},
        {5,6,7,8},
        {9,10,11,12},
        {13,14,15,16}};

        SpiralMatrix(matrix);
           
    }
    public static void SpiralMatrix(int matrix[][]){
        int startrow =0;
        int endrow= matrix.length-1;
        int startcols = 0;
        int endcols = matrix[0].length-1;

        while (startcols<=endcols && startrow<=endcols) {
            //top
            for(int j=startcols; j<=endcols; j++){
                System.out.print(matrix[startrow][j] + " ");
            }

            //Right
            for(int i = startrow+1; i<=endrow; i++){
                System.out.print(matrix[i][endcols] +" ");
            }

            // left
            for(int j=endcols-1; j>=startcols; j--){
                System.out.print(matrix[endrow][j] +" ");
            }

             //up
             for(int j=endrow-1; j>=startrow+1; j--){
                System.out.print(matrix[j][startcols] +" ");
            }

            startrow++;
            endrow--;
            startcols++;
            endcols--;




        }
    }
}

File: Stiver DSA Sheet/Recursion/spiral Matrix.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;

void spiralOrder(vector<vector<int>> &mat)
{
    int top = 0, left = 0;
    int bottom = mat.size() - 1;
    int right = mat[0].size() - 1;

    while (top <= bottom && left <= right)
    {
        for (int i = left; i <= right; i++)
        {
            cout << mat[top][i] << " ";
        }
        top++;

        for (int i = top; i <= bottom; i++)
        {
            cout << mat[i][right] << " ";
        }
        right--;

        if (top <= bottom)
        {
            for (int i = right; i >= left; i--)
            {
                cout << mat[bottom][i] << " ";
            }
            bottom--;
        }

        if (left <= right)
        {
            for (int i = bottom; i >=top; i--)
            {
                cout << mat[i][left] << " ";
            }
            left++;
        }
    }

}

int countFactor(int n ){
    int cnt =0;
    for (int i = 1; i *i  < n; i++)
    {
       if(n % i == 0){
         cnt += (i*i == n)?1:2;
       }
    }
    return cnt;
}

string removeDuplicate(string &a){
    string res;
    for(char c: a){
        if (a.empty() || a.back() != c)
        {
            res+=c;
        }
        
    }
    return res;
}

int main()
{
    vector<vector<int>> matrix = {
        {1, 2, 3, 4},
        {5, 6, 7, 8},
        {9, 10, 11, 12}};
    
    vector<int> ans = {10, 15, 21};
    vector<int> res;
    for(int num : ans){
        res.push_back(countFactor(num));
    }
    for(int ele :  res){
        cout<<ele<<" ";
    }
    // spiralOrder(matrix);
    return 0;
}